sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -394,4 +394,72 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       Seq(Row(true), Row(true))
     )
   }
+
+  test("time windowing") {
+    // 2016-03-27 19:39:34 UTC, 2016-03-27 19:39:56 UTC, 2016-03-27 19:39:27 UTC
+    val df = Seq((1459103974L, 1, "a"), (1459103996L, 2, "a"),
+      (1459103967L, 4, "b")).toDF("time", "value", "id")
+    checkAnswer(
+      df.groupBy(window($"time", "10 seconds")).agg(count("*").as("counts"))
+        .orderBy($"window.start".asc).select("counts"),
+      Seq(Row(1), Row(1), Row(1))
+    )
+    checkAnswer(
+      df.groupBy(window($"time", "10 seconds", "10 seconds", "5 seconds"))
+        .agg(count("*").as("counts")).orderBy($"window.start".asc).select("counts"),
+      Seq(Row(2), Row(1))
+    )
+    checkAnswer(
+      df.groupBy(window($"time", "10 seconds", "10 seconds", "5 seconds"), $"id")
+        .agg(count("*").as("counts")).orderBy($"window.start".asc).select("counts"),
+      Seq(Row(1), Row(1), Row(1))
+    )
+    checkAnswer(
+      df.groupBy(window($"time", "10 seconds", "3 seconds", "0 second"))
+        .agg(count("*").as("counts")).orderBy($"window.start".asc).select("counts"),
+      // 2016-03-27 19:39:27 UTC -> 4 bins
+      // 2016-03-27 19:39:34 UTC -> 3 bins
+      // 2016-03-27 19:39:56 UTC -> 3 bins
+      Seq(Row(1), Row(1), Row(1), Row(2), Row(1), Row(1), Row(1), Row(1), Row(1))
+    )
+    checkAnswer(
+      df.select(window($"time", "10 seconds"), $"value")
+        .orderBy($"window.start".asc).select("value"),
+      Seq(Row(4), Row(1), Row(2))
```

Comment:
It might be helpful to separate these into tests named `tumbling`, `overlapping` etc.  Just to make it clearer that we are covering all the cases well.